network/p2p/acp118/aggregator.go

```diff
@@ -0,0 +1,254 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package acp118
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p"
+	"github.com/ava-labs/avalanchego/proto/pb/sdk"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp"
+)
+
+// ErrFailedAggregation is returned if it's not possible for us to
+// generate a signature
+var (
+	ErrFailedAggregation  = errors.New("failed aggregation")
+	errFailedVerification = errors.New("failed verification")
+)
+
+// Validator signs warp messages. NodeID must be unique across validators, but
+// PublicKey is not guaranteed to be unique.
+type Validator struct {
+	NodeID    ids.NodeID
+	PublicKey *bls.PublicKey
+	Weight    uint64
+}
+
+type indexedValidator struct {
+	Validator
+	Index int
+}
+
+type result struct {
+	Validator indexedValidator
+	Signature *bls.Signature
+	Err       error
+}
+
+// NewSignatureAggregator returns an instance of SignatureAggregator
+func NewSignatureAggregator(log logging.Logger, client *p2p.Client) *SignatureAggregator {
+	return &SignatureAggregator{
+		log:    log,
+		client: client,
+	}
+}
+
+// SignatureAggregator aggregates validator signatures for warp messages
+type SignatureAggregator struct {
+	log    logging.Logger
+	client *p2p.Client
+}
+
+// AggregateSignatures blocks until quorumNum/quorumDen signatures from
+// validators are requested to be aggregated into a warp message or the context
+// is canceled. Returns the signed message and the amount of stake that signed
+// the message. Caller is responsible for providing a well-formed canonical
+// validator set corresponding to the signer bitset in the message.
+func (s *SignatureAggregator) AggregateSignatures(
+	ctx context.Context,
+	message *warp.Message,
+	justification []byte,
+	validators []Validator,
+	quorumNum uint64,
+	quorumDen uint64,
+) (*warp.Message, uint64, uint64, error) {
```

Comment:
I don't think this correctly handles the case that BLS public keys are shared across validators.

In Warp, only 1 signature is ever allowed from a BLS key in a warp message. If different nodeIDs have the same BLS key, their weights are aggregated for the BLS key's index

Reply:
Updated to use the `warp.Validator` type + added a test on shared public keys